ledger-service/lf-value-json/specification.rst

```diff
@@ -0,0 +1,434 @@
+    { foo: 42 }      -->  Depth1 { foo: Some 42 }         :  Depth1
+    { foo: [42] }    -->  Depth2 { foo: Some (Some 42) }  :  Depth2
+    { foo: null }    -->  Depth1 { foo: None }            :  Depth1
+    { foo: null }    -->  Depth2 { foo: None }            :  Depth2
+    { foo: [] }      -->  Depth2 { foo: Some None }       :  Depth2
+
+Note that the shortcut for records and Optional fields does not apply to
+Map (which are also represented as JSON objects), since Map relies on
+absence of key to determine what keys are present in the Map to begin
+with.  Nor does it apply to the ``[f₁, ..., fₙ]`` record form; ``Depth1
+None`` in the array notation must be written as ``[null]``.
+
+Type variables may appear in the DAML-LF language, but are always
+resolved before deciding on a JSON encoding.  So, for example, even
+though ``Oa`` doesn't appear to contain a nested ``Optional``, it may
+contain a nested ``Optional`` by virtue of substituting the type
+variable ``a``::
+
+    record Oa a = { foo: Optional a }
+
+    { foo: 42 }     -->  Oa { foo: Some 42 }        : Oa Int
+    { }             -->  Oa { foo: None }           : Oa Int
+    { foo: [] }     -->  Oa { foo: Some None }      : Oa (Optional Int)
+    { foo: [42] }   -->  Oa { foo: Some (Some 42) } : Oa (Optional Int)
+
+In other words, the correct JSON encoding for any LF value is the one
+you get when you have eliminated all type variables.
+
+Output
+~~~~~~
+
+Encoded as described above, always applying the shortcut for None record
+fields.
+
+Variant
+-------
+
+Variants are expressed as
+
+::
+
+    { constructor: argument }
+
+For example, if we have
+
+::
+
+    variant Foo = Bar Int64 | Baz Unit | Quux (Optional Int64)
+
+These are all valid JSON encodings for values of type Foo::
+
+    {"Bar": 42}
+    {"Baz": {}}
+    {"Quux": null}
+    {"Quux": 42}
+
+Note that DAML data types with named fields are compiled by factoring
+out the record. So for example if we have
+
+::
+
+    data Foo = Bar {f1: Int64, f2: Bool} | Baz
+
+We'll get in DAML-LF
+
+::
+
+    record Foo.Bar = {f1: Int64, f2: Bool}
+    variant Foo = Bar Foo.Bar | Baz Unit
+
+and then, from JSON
+
+::
+
+    {"Bar": {"f1": 42, "f2": true}}
+    {"Baz": {}}
+
+This can be encoded and used in TypeScript, including exhaustiveness
+checking; see `a keyed example`_.
+
+.. _a keyed example: https://www.typescriptlang.org/play/#src=type%20Foo%20%3D%0D%0A%20%20%20%20%7B%20Bar%3A%20%7B%20f1%3A%20number%2C%20f2%3A%20boolean%20%7D%20%7D%0D%0A%20%20%7C%20%7B%20Baz%3A%20%7B%20f3%3A%20string%20%7D%20%7D%3B%0D%0A%0D%0Afunction%20test(v%3A%20Foo)%20%7B%0D%0A%20%20if%20(%22Bar%22%20in%20v)%20%7B%0D%0A%20%20%20%20console.log(v.Bar.f1%2C%20v.Bar.f2)%3B%0D%0A%20%20%7D%20else%20if%20(%22Baz%22%20in%20v)%20%7B%0D%0A%20%20%20%20console.log(v.Baz.f3)%3B%0D%0A%20%20%7D%20else%20%7B%0D%0A%20%20%20%20const%20_%3A%20never%20%3D%20v%3B%0D%0A%20%20%7D%0D%0A%7D%20%0D%0A
+
+Enum
+----
+
+Enums are represented as strings. So if we have
```

Comment:
Enums are represented as JSON strings.

Reply:
Is that not clear?

Reply:
you are saying above that records represented as JSON objects.... so for consistency, I would say as JSON strings here too.

Reply:
I'll delete the JSON prefix from objects.